gpflow/conditionals/util.py

```diff
@@ -373,38 +445,56 @@ def rollaxis_right(A, num_rolls):
     return tf.transpose(A, perm)
 
 
-def mix_latent_gp(W, g_mu, g_var, full_cov, full_output_cov):
+def mix_latent_gp(W, g_mean, g_var, full_cov, full_output_cov):
     r"""Takes the mean and variance of an uncorrelated L-dimensional latent GP
     and returns the mean and the variance of the mixed GP, `f = W g`,
     where both f and g are GPs, with W having a shape [P, L]
 
     :param W: [P, L]
-    :param g_mu: [..., N, L]
+    :param g_mean: [..., N, L]
     :param g_var: [..., N, L] (full_cov = False) or [L, ..., N, N] (full_cov = True)
-    :return: f_mu and f_var, shape depends on `full_cov` and `full_output_cov`
+    :return: f_mean and f_var, shape depends on `full_cov` and `full_output_cov`
     """
-    f_mu = tf.tensordot(g_mu, W, [[-1], [-1]])  # [..., N, P]
+    shape_constraints = [
+        (W, "PL"),
+        (g_mean, "*NL"),
+    ]
+    if not full_cov:
+        shape_constraints.append((g_var, "*NL"))
```

Comment:
why would g_var not have constraints if not full_cov?

Reply:
because we can't check if there are inner "leading" dimensions (like above, see https://github.com/GPflow/GPflow/issues/1296). in one case, g_var is rolled, and the shape_constraint is added below

Reply:
(I've added a comment to clarify)